9290 device removal reduces redundancy of mirrors

ahrens opened the pull request:
Mirrors are supposed to provide redundancy in the face of whole-disk
failure and silent damage (e.g. some data on disk is not right, but ZFS
hasn't detected the whole device as being broken). However, the current
device removal implementation bypasses some of the mirror's redundancy.
Note that in no case is incorrect data returned, but we might get a
checksum error when we should have been able to find the right data.

There are two underlying problems:

1. When we remove a mirror device, we only read one side of the mirror.
Since we can't verify the checksum, this side may be silently bad, but
the good data is on the other side of the mirror (which we didn't read).
This can cause the removal to "bake in" the busted data – all copies of
the data in the new location are the same, busted version, while we left
the good version behind.

The fix for this is to read and copy both sides of the mirror. If the
old and new vdevs are mirrors, we will read both sides of the old
mirror, and write each copy to the corresponding side of the new mirror.
(If the old and new vdevs have a different number of children, we will
do this as best as possible.) Even though we aren't verifying checksums,
this ensures that as long as there's a good copy of the data, we'll have
a good copy after the removal, even if there's silent damage to one side
of the mirror. If we're removing a mirror that has some silent damage,
we'll have exactly the same damage in the new location (assuming that
the new location is also a mirror).

2. When we read from an indirect vdev that points to a mirror vdev, we
only consider one copy of the data. This can lead to reduced effective
redundancy, because we might read a bad copy of the data from one side
of the mirror, and not retry the other, good side of the mirror.

Note that the problem is not with the removal process, but rather after
the removal has completed (having copied correct data to both sides of
the mirror), if one side of the new mirror is silently damaged, we
encounter the problem when reading the relocated data via the indirect
vdev. Also note that the problem doesn't occur when ZFS knows that one
side of the mirror is bad, e.g. when a disk entirely fails or is
offlined.

The impact is that reads (from indirect vdevs that point to mirrors) may
return a checksum error even though the good data exists on one side of
the mirror, and scrub doesn't repair all data on the mirror (if some of
it is pointed to via an indirect vdev).

The fix for this is complicated by "split blocks" - one logical block
may be split into two (or more) pieces with each piece moved to a
different new location. In this case we need to read all versions of
each split (one from each side of the mirror), and figure out which
combination of versions results in the correct checksum, and then repair
the incorrect versions.

This ensures that we supply the same redundancy whether you use device
removal or not. For example, if a mirror has small silent errors on all
of its children, we can still reconstruct the correct data, as long as
those errors are at sufficiently-separated offsets (specifically,
separated by the largest block size - default of 128KB, but up to 16MB).

ahrens committed:
9290 device removal reduces redundancy of mirrors

Mirrors are supposed to provide redundancy in the face of whole-disk
failure and silent damage (e.g. some data on disk is not right, but ZFS
hasn't detected the whole device as being broken). However, the current
device removal implementation bypasses some of the mirror's redundancy.
Note that in no case is incorrect data returned, but we might get a
checksum error when we should have been able to find the right data.

There are two underlying problems:

1. When we remove a mirror device, we only read one side of the mirror.
Since we can't verify the checksum, this side may be silently bad, but
the good data is on the other side of the mirror (which we didn't read).
This can cause the removal to "bake in" the busted data – all copies of
the data in the new location are the same, busted version, while we left
the good version behind.

The fix for this is to read and copy both sides of the mirror. If the
old and new vdevs are mirrors, we will read both sides of the old
mirror, and write each copy to the corresponding side of the new mirror.
(If the old and new vdevs have a different number of children, we will
do this as best as possible.) Even though we aren't verifying checksums,
this ensures that as long as there's a good copy of the data, we'll have
a good copy after the removal, even if there's silent damage to one side
of the mirror. If we're removing a mirror that has some silent damage,
we'll have exactly the same damage in the new location (assuming that
the new location is also a mirror).

2. When we read from an indirect vdev that points to a mirror vdev, we
only consider one copy of the data. This can lead to reduced effective
redundancy, because we might read a bad copy of the data from one side
of the mirror, and not retry the other, good side of the mirror.

Note that the problem is not with the removal process, but rather after
the removal has completed (having copied correct data to both sides of
the mirror), if one side of the new mirror is silently damaged, we
encounter the problem when reading the relocated data via the indirect
vdev. Also note that the problem doesn't occur when ZFS knows that one
side of the mirror is bad, e.g. when a disk entirely fails or is
offlined.

The impact is that reads (from indirect vdevs that point to mirrors) may
return a checksum error even though the good data exists on one side of
the mirror, and scrub doesn't repair all data on the mirror (if some of
it is pointed to via an indirect vdev).

The fix for this is complicated by "split blocks" - one logical block
may be split into two (or more) pieces with each piece moved to a
different new location. In this case we need to read all versions of
each split (one from each side of the mirror), and figure out which
combination of versions results in the correct checksum, and then repair
the incorrect versions.

This ensures that we supply the same redundancy whether you use device
removal or not. For example, if a mirror has small silent errors on all
of its children, we can still reconstruct the correct data, as long as
those errors are at sufficiently-separated offsets (specifically,
separated by the largest block size - default of 128KB, but up to 16MB).

ahrens commented:
@allanjude @dweeezil Thanks for looking at the previous iteration of this.  The new and improved fix is much more comprehensive but the code is almost entirely different.  Could I ask that you take a look at this too?

ahrens committed:
no repair when readonly

Reviewed at: http://reviews.delphix.com/r/38497/

dweeezil commented:
@ahrens I'll look it over and try to get this into the ZoL device evacuation port this weekend.

dweeezil commented on usr/src/uts/common/fs/zfs/vdev_indirect.c:
```diff
@@ -1065,41 +1136,438 @@ vdev_indirect_child_io_done(zio_t *zio)
+		ret = zio_checksum_error(zio, &zbc);
+		if (ret == 0) {
+			/* Found a matching checksum.  Issue repair i/os. */
+			vdev_indirect_repair(zio);
```
I added this here:
```
diff --git a/module/zfs/vdev_indirect.c b/module/zfs/vdev_indirect.c
index f794b92..7d88b4c 100644
--- a/module/zfs/vdev_indirect.c
+++ b/module/zfs/vdev_indirect.c
@@ -1416,7 +1416,8 @@ vdev_indirect_reconstruct_io_done(zio_t *zio)
                ret = zio_checksum_error(zio, &zbc);
                if (ret == 0) {
                        /* Found a matching checksum.  Issue repair i/os. */
-                       vdev_indirect_repair(zio);
+                       if (spa_writeable(zio->io_spa))
+                               vdev_indirect_repair(zio);
                        zio_checksum_verified(zio);
                        return;
                }
```
to keep zdb from triggering ```VERIFY(zio->io_type != ZIO_TYPE_WRITE || spa_writeable(spa));``` in ```zio_vdev_io_start()```.

ahrens replied:
Sorry about that, I found that problem too but forgot to update the PR.  I fixed it in a slightly different, but equivalent way.

dweeezil reviewed (commented):
So far, looks good (in the port to ZoL) except for the repair writes happening for read-only pools.

ahrens committed:
DLPX-57626 device removal doesn't tolerate concurrent vdev config changes

ahrens commented:
I've added another commit to this which addresses an additional problem with device removal and mirrors:

Now that we support device removal with mirrors, it's reasonable to change the pool config while a removal is in progress. E.g. to "zpool attach", "zpool detach", or automatic spare in / out events. However, these operations cause panics because we save the vdev_t that is being removed in svr_vdev, but this vdev_t can be freed (e.g. if a detach changes the top-level vdev from a mirror to a plain disk).

The solution is to store the vdev ID in the spa_vdev_removal_t, rather than the vdev_t. Most places that the svr_vdev is used will have to use vdev_lookup_top() to obtain the vdev_t (in some cases adding additional calls to spa_config_enter/exit).

ahrens committed:
post event

behlendorf commented on usr/src/uts/common/fs/zfs/vdev_indirect.c:
```diff
@@ -1065,41 +1136,441 @@ vdev_indirect_child_io_done(zio_t *zio)
+		/* copy data from splits to main zio */
+		int ret;
+		for (indirect_split_t *is = list_head(&iv->iv_splits);
+		    is != NULL; is = list_next(&iv->iv_splits, is)) {
```
When testing with `zloop.sh` I observed that when there are a large number of segment copies and the damaged copy is one of the "high bits" then this function can't be expected to check everything.  This manifested itself as a `zdb` hang in my case and a few suspended pools due to stalling the IO pipeline.

I'm proposing https://github.com/behlendorf/zfs/commit/2ad01af1b4a50117779aab7e47c7006a7f2451e6 as a fix . It slightly reworks the logic in `vdev_indirect_reconstruct_io_done` to adopt a random sampling apprach to reconstruction when there are a large number of segments, otherwise we use the existing logic and check everything.  We could further optimize this if we decide it's needed.

ahrens committed:
Fix zdb hang when running zloop

Reviewed at: http://reviews.delphix.com/r/38497/

ahrens committed:
serapheim feedback

Reviewed at: http://reviews.delphix.com/r/38497/

ahrens commented:
I added 2 more commits, which address feedback from @sdimitro and @behlendorf.

behlendorf reviewed (approved):
I like it, thanks.

dweeezil committed:
Don't clear ranges beyond end of spacemap

Signed-off-by: Tim Chase <tim@chase2k.com>
Requires-builders: none

ahrens committed:
comments

Reviewed at: http://reviews.delphix.com/r/38497/

ahrens committed:
removing vdev can be resilvered

Reviewed at: http://reviews.delphix.com/r/38497/

ahrens commented:
Some minor changes to address feedback from @grwilson, @prashks, and Sara Hartse.

behlendorf committed:
Reasonably limit fault injection mayhem

If there is a vdev removal in progress, we can't risk causing
additional damage to newly evacuated blocks which might not
be fully redundant due to a previous fault injection.

Signed-off-by: Brian Behlendorf <behlendorf1@llnl.gov>
Requires-builders: none

ahrens committed:
fix lint

ahrens committed:
nits

ahrens commented:
I think we can also count @dweeezil and @behlendorf as reviewers.
